autoload/lsp/ui/vim.vim

```diff
@@ -371,36 +430,201 @@ function! s:apply_workspace_edits(workspace_edits) abort
     endif
 endfunction
 
-function! s:generate_move_cmd(line_pos, character_pos) abort
-    let l:result = printf("%dG0", a:line_pos) " move the line and set to the cursor at the beggining
-    if a:character_pos > 0
-        let l:result .= printf("%dl", a:character_pos) " Move right until the character
+function! s:apply_text_edits(uri, text_edits) abort
+    " https://microsoft.github.io/language-server-protocol/specification#textedit
+    " The order in the array defines the order in which the inserted string
+    " appear in the resulting text.
+    "
+    " The edits must be applied in the reverse order so the early edits will
+    " not interfere with the position of later edits, they need to be applied
+    " one at the time or put together as a single command.
+    "
+    " Example: {"range": {"end": {"character": 45, "line": 5}, "start":
+    " {"character": 45, "line": 5}}, "newText": "\n"}, {"range": {"end":
+    " {"character": 45, "line": 5}, "start": {"character": 45, "line": 5}},
+    " "newText": "import javax.ws.rs.Consumes;"}]}}
+    "
+    " If we apply the \n first we will need adjust the line range of the next
+    " command (so the import will be written on the next line) , but if we
+    " write the import first and then the \n everything will be fine.
+    " If you do not apply a command one at  time, you will need to adjust the
+    " range columns after which edit. You will get this (only one execution):
+    "
+    " execute 'keepjumps normal! 6G045laimport javax.ws.rs.Consumes;'" |
+    " execute 'keepjumps normal! 6G045la\n'
+    "
+    " resulting in this:
+    " import javax.servlet.http.HttpServletRequest;i
+    " mport javax.ws.rs.Consumes;
+    "
+    " instead of this (multiple executions):
+    " execute 'keepjumps normal! 6G045laimport javax.ws.rs.Consumes;'"
+    " execute 'keepjumps normal! 6G045li\n'
+    "
+    " resulting in this:
+    " import javax.servlet.http.HttpServletRequest;
+    " import javax.ws.rs.Consumes;
+    "
+    "
+    " The sort is also necessary since the LSP specification does not
+    " guarantee that text edits are sorted.
+    "
+    " Example:
+    " Initial text:  "abcdef"
+    " Edits:
+    " ((0,0), (0, 1), "") - remove first character 'a'
+    " ((0, 4), (0, 5), "") - remove fifth character 'e'
+    " ((0, 2), (0, 3), "") - remove third character 'c'
+    for l:text_edit in reverse(s:stable_sort(a:text_edits, function('s:compare_text_edits')))
```

Comment:
can skip the `reverse` if you swap the operands when you do the subtraction in the comparison function.